frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -1305,6 +1305,114 @@ const IR::Node* TypeInference::postorder(IR::Concat* expression) {
     return expression;
 }
 
+const IR::Node* TypeInference::postorder(IR::Key* key) {
+    LOG1("TypeInference: visiting " << dbp(key));
+    // compute the type and store it in typeMap
+    auto vec = new IR::Vector<IR::Type>();
+    for (auto ke : *key->keyElements) {
+        auto kt = typeMap->getType(ke->expression);
+        if (kt == nullptr) {
+            LOG2("Bailing out for " << dbp(ke));
+            return key;
+        }
+        vec->push_back(kt);
+    }
+    auto keyTuple = new IR::Type_Tuple(Util::SourceInfo(), vec);
+    LOG2("Setting key type to " << dbp(keyTuple));
+    setType(key, keyTuple);
+    // installing also for the original because we cannot tell which one will survive in the ir
+    // \TODO: figure out the rules of survivorship
```

Comment:
Note that if you really want to know whether this node has changed in a postorder function, and only insert into the typemap in that case, you can use:

    if (*key != *getOriginal())
        setType(key, keyTuple);

This doesn't work in a preorder function as it only tells you if the ndoe has changed yet, and not if it might be changed in the future from traversing its children or in the postorder function.